File: GPACalculator.java
import java.util.Scanner;

class GPACalculator {

  public static void main(String[] args) {
		
		Scanner input = new Scanner (System.in);
		
		System.out.println("Welcome to the GPA Calculator!");
		System.out.println("How many courses do you take?");
		int courses = input.nextInt();

		int counter = 1;
		String[] grade = new String[courses];
		char[] level = new char[courses];
		double[] credits = new double[courses];
		
		while (counter <= courses)
		{
			System.out.println("Please enter the grade for course " + counter);
			grade[counter - 1] = input.next();
			System.out.println("Please enter the credits for course " + counter);
			credits[counter - 1] = input.nextDouble();
			System.out.println("Please enter the level (H = honors, A = accelerated, R = regular) for course " + counter);
			level[counter - 1] = input.next().charAt(0);
			counter++;
		}
		input.close();
		
		double unweighted_gpa = uGPA(grade, level, credits, courses);
		double weighted_gpa = wGPA(grade, level, credits, courses);
		
		System.out.println("Your unweighted GPA is: " + unweighted_gpa);
		System.out.println("Your weighted GPA is: " + weighted_gpa);
	}

	public static double uGPA (String[] grade, char[] level, double[] credits, int courses)
	{
		int counter = 1;
		double unweighted_gpa, totalvalue = 0, totalcredits = 0;
		double[] value = new double[courses];
		
		while (counter <= courses)
		{
			value[counter - 1] = getValue(counter, grade);

			totalvalue += credits[counter - 1] * value[counter - 1];
			totalcredits += credits[counter - 1];
			counter++;
		}
		unweighted_gpa = totalvalue/totalcredits;
		
		return unweighted_gpa;
	}
	
	public static double wGPA (String[] grade, char[] level, double[] credits, int courses)
	{
		int counter = 1;
		double weighted_gpa, totalvalue = 0, totalcredits = 0;
		double[] value = new double[courses];
		
		while (counter<=courses)
		{
			value[counter - 1] = getValue(counter, grade);
			if (level[counter - 1] == 'A' && grade[counter - 1] != "D" && grade[counter - 1] != "F")
			{
				value[counter - 1] += 1;
			}
			if (level[counter - 1] == 'H' && grade[counter - 1] != "D" && grade[counter - 1] != "F")
			{
			  value[counter - 1] += 2;
			}
				  
			totalvalue += credits[counter - 1] * value[counter - 1];
			totalcredits += credits[counter - 1];
			counter++;
		}
		weighted_gpa = totalvalue/totalcredits;
				
		return weighted_gpa;
	}

	private static double getValue (int counter, String[] grade)
	{
		switch (grade[counter - 1].toUpperCase()) {
			case "A+": return 4.33;
			case "A": return 4.00;
			case "A-": return 3.67;
			case "B+": return 3.33;
			case "B": return 3.00;
			case "B-": return 2.67;
			case "C+": return 2.33;
			case "C": return 2.00;
			case "C-": return 1.67;
			case "D": return 1.00;
			default: return 0.00;
		}
  }
  
}

File: gpa_calculator.py
def uGPA(grds, crds):
    totalvalue, totalcredits = 0, 0
    for c in range(len(grds)):
        totalvalue += crds[c] * getValue(c, grds)
        totalcredits += crds[c]
    return totalvalue / totalcredits


def wGPA(grds, lvls, crds):
    totalvalue, totalcredits = 0, 0
    for c in range(len(grds)):
        value = getValue(c, grds)
        if lvls[c] == 'A' and grds[c] != 'D' and 'F':
            value += 1
        if lvls[c] == 'H' and grds[c] != 'D' and 'F':
            value += 2
        totalvalue += crds[c] * value
        totalcredits += crds[c]
    return totalvalue / totalcredits


def getValue(num, grds):
    grade = grds[num].upper()
    if grade == 'A+':
        return 4.33
    elif grade == 'A':
        return 4.00
    elif grade == 'A-':
        return 3.67
    elif grade == 'B+':
        return 3.33
    elif grade == 'B':
        return 3.00
    elif grade == 'B-':
        return 2.67
    elif grade == 'C+':
        return 2.33
    elif grade == 'C':
        return 2.00
    elif grade == 'C-':
        return 1.67
    elif grade == 'D':
        return 1.00
    else:
        return 0.00


try:
    courses = int(input('Welcome to the GPA Calculator!\nHow many courses do you take?\n'))
except:
    quit('The input is not an integer. Please try again.')
if courses <= 0:
    quit('The number of courses must be greater than zero. Please try again.')

grades, levels, creds = [], [], []

for i in range(courses):
    grades.append(input('Please enter the grade for course ' + str(i + 1) + '\n'))
    if grades[-1].upper() not in ['A+', 'A', 'A-', 'B+', 'B', 'B-', 'C+', 'C', 'C-', 'D', 'F']:
        quit('The entered grade is not valid. Please try again.')
    try:
        creds.append(float(input('Please enter the credits for course ' + str(i + 1) + '\n')))
    except:
        quit('The input is not a floating point number. Please try again.')
    if creds[-1] <= 0:
        quit('The number of credits must be greater than zero. Please try again.')
    levels.append(input(
        'Please enter the level (H = Honors/AP, A = Accelerated, R = Regular) for course ' + str(i + 1) + '\n'))
    if levels[-1].upper() not in ['H', 'A', 'R']:
        quit('The entered level is not valid. Please try again.')

print('Your unweighted GPA is:', uGPA(grades, creds), '\nYour weighted GPA is:', wGPA(grades, levels, creds))
